File: solutions/javascript/bfs/numberOfProvinces.js
const CONNECTED = 1;

const markProvinceAsVisited = (startCity, cityConnections, visited) => {
    const numCities = cityConnections.length;

    const queue = new Queue();
    queue.enqueue(startCity);

    while (queue.size() > 0) {
        // Remove node
        const city = queue.dequeue();

        // Process node
        visited.add(city);

        // Add neighbors
        for (
            let potentialNeighborCity = 0;
            potentialNeighborCity < numCities;
            potentialNeighborCity++
        ) {
            const isNeighbor =
                cityConnections[city][potentialNeighborCity] === CONNECTED;
            if (!isNeighbor || visited.has(potentialNeighborCity)) continue;

            queue.enqueue(potentialNeighborCity);
        }
    }
};

const findCircleNum = (cityConnections) => {
    const numCities = cityConnections.length;

    let numProvinces = 0;
    const visited = new Set();

    for (let city = 0; city < numCities; city++) {
        if (visited.has(city)) continue;

        numProvinces++;
        markProvinceAsVisited(city, cityConnections, visited);
    }

    return numProvinces;
};

File: solutions/python/bfs/number_of_provinces.py
from collections import deque

CONNECTED = 1

class Solution:
    def get_neighbors(self, node_id, adjacency_matrix):
        num_nodes = len(adjacency_matrix)
        neighbors = []

        for potential_neighbor_id in range(num_nodes):
            is_neighbor = adjacency_matrix[node_id][potential_neighbor_id] == CONNECTED
            if not is_neighbor:
                continue

            neighbors.append(potential_neighbor_id)

        return neighbors

    def mark_component_as_visited(self, start_node_id, adjacency_matrix, visited):
        queue = deque()
        queue.append(start_node_id)

        while queue:
            # Remove node
            node_id = queue.popleft()

            # Process node
            # No work required here

            # Add neighbors
            neighbor_ids = self.get_neighbors(node_id, adjacency_matrix)
            for neighbor_id in neighbor_ids:
                if neighbor_id in visited:
                    continue

                visited.add(neighbor_id)
                queue.append(neighbor_id)
            

    def findCircleNum(self, cityConnections: List[List[int]]) -> int:
        num_cities = len(cityConnections)
        num_provinces = 0

        visited = set()

        for city in range(num_cities):
            if city in visited:
                continue

            num_provinces += 1
            self.mark_component_as_visited(city, cityConnections, visited)
            
        return num_provinces

